server/lib/scheduler/UserSchedule.ts

```diff
@@ -3,50 +3,84 @@ import { Assignment } from "./Assignment";
 import { RootParentFinder } from "./RootParentFinder";
 
 class UserSchedule {
+  private lastWorkedOnTimestampForTree = {};
+  private rootParentCache;
   private timeLimit;
   private userId;
   private userSchedule = [];
 
-  public constructor({ timeLimit, userId }) {
+  public constructor({ rootParentCache, timeLimit, userId }) {
+    this.rootParentCache = rootParentCache;
     this.timeLimit = timeLimit;
     this.userId = userId;
   }
 
-  public hasUserBeenAssignedToAnyWorkspaces() {
-    return this.userSchedule.length > 0;
-  }
-
-  public assignWorkspace(workspace, startAtTimestamp = Date.now()) {
+  public async assignWorkspace(workspace, startAtTimestamp = Date.now()) {
     const assignment = new Assignment({
       userId: this.userId,
       workspace,
       startAtTimestamp,
     });
+
     this.userSchedule.push(assignment);
+
+    const rootParent = await this.rootParentCache.getRootParentOfWorkspace(workspace);
+
+    this.lastWorkedOnTimestampForTree[rootParent.id] = startAtTimestamp;
   }
 
   public getMostRecentAssignment() {
     return this.userSchedule[this.userSchedule.length - 1];
   }
 
+  public getTreesWorkedOnLeastRecentlyByUser(rootWorkspaces) {
+    const treesNotYetWorkedOn = rootWorkspaces.filter(
+      r => this.lastWorkedOnTimestampForTree[r.id] === undefined
+    );
```

Comment:
this feels like it should be a utility method, like,
``public hasNotBeenWorkedOn(workspaceId)``

But maybe in a different file or similar.